libs/vkd3d/raytracing_pipeline.c

```diff
@@ -0,0 +1,778 @@
+/*
+ * Copyright 2021 Hans-Kristian Arntzen for Valve Corporation
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#define VKD3D_DBG_CHANNEL VKD3D_DBG_CHANNEL_API
+#include "vkd3d_private.h"
+#include "vkd3d_string.h"
+
+static inline struct d3d12_state_object *impl_from_ID3D12StateObject(ID3D12StateObject *iface)
+{
+    return CONTAINING_RECORD(iface, struct d3d12_state_object, ID3D12StateObject_iface);
+}
+
+static inline struct d3d12_state_object *impl_from_ID3D12StateObjectProperties(ID3D12StateObjectProperties *iface)
+{
+    return CONTAINING_RECORD(iface, struct d3d12_state_object, ID3D12StateObjectProperties_iface);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_state_object_QueryInterface(ID3D12StateObject *iface,
+        REFIID riid, void **object)
+{
+    TRACE("iface %p, riid %s, object %p.\n", iface, debugstr_guid(riid), object);
+
+    if (IsEqualGUID(riid, &IID_ID3D12StateObject)
+        || IsEqualGUID(riid, &IID_ID3D12DeviceChild)
+        || IsEqualGUID(riid, &IID_ID3D12Object)
+        || IsEqualGUID(riid, &IID_IUnknown))
+    {
+        ID3D12StateObject_AddRef(iface);
+        *object = iface;
+        return S_OK;
+    }
+
+    if (IsEqualGUID(riid, &IID_ID3D12StateObjectProperties))
+    {
+        struct d3d12_state_object *state_object = impl_from_ID3D12StateObject(iface);
+        ID3D12StateObjectProperties_AddRef(&state_object->ID3D12StateObjectProperties_iface);
+        *object = &state_object->ID3D12StateObjectProperties_iface;
+        return S_OK;
+    }
+
+    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(riid));
+
+    *object = NULL;
+    return E_NOINTERFACE;
+}
+
+static HRESULT STDMETHODCALLTYPE d3d12_state_object_properties_QueryInterface(ID3D12StateObjectProperties *iface,
+        REFIID riid, void **object)
+{
+    TRACE("iface %p, riid %s, object %p.\n", iface, debugstr_guid(riid), object);
+
+    if (IsEqualGUID(riid, &IID_ID3D12StateObjectProperties)
```

Comment:
Going by COM rules, this should call `d3d12_state_object_QueryInterface` with the parent state object.

May need some testing on Windows, but in general, if interface A allows querying interface B, then interface B must also allow querying interface A and pointers to IUnknown instances have to be consistent.

Reply:
Fixed.